### PR TITLE
tests: new mkimage-uc22 tool

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This new tool is used to build a classic ubuntu image with kernel, gadget snapd and base snaps.

This script is based on the one stored in the
tests/nested/manual/fde-on-classic.

This change also includes a new spread test done to validate the tool.

The new tool has new parameters which make the process easy to run in different scenarios.

### CODE REVIEW

File: tests/lib/tools/remote.wait-for [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# This is a placeholder for the tool which is copied from snapd-testing-tools project
+# The tool is copied from the tests/lib/external/snapd-testing-tools/remote directory
+# while the project is being prepared
+echo "placeholder called, source tree not properly setup"
+exit 1
```
Comment (Meulengracht) [inline tests/lib/tools/remote.wait-for:5]: missing newline?

File: tests/lib/tools/mkimage-uc22 [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+        fi
+    done
+
+    mkdir -p "$CACHE"
+    # shellcheck disable=SC2086
+    cp $SNAPS "$CACHE"
+
+    # get the snaps
+    for snap in pc-kernel pc; do        
+        NAME="$(find_snap "$CACHE" "$snap")"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:67]: could NAME be local?

File: tests/lib/tools/mkimage-uc22 [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+        echo "mkimage-uc22: downloading the ubuntu classic base"
+        wget -c -P "$CACHE" "$BASE_IMAGE_URL"
+    fi    
+
+    echo "mkimage-uc22: image assets ready"
+}
+
+# shellcheck disable=SC2317
+cleanup() {
+    IMG="$(readlink -f "$1")"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:99]: local variables here?

File: tests/lib/tools/mkimage-uc22 [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+    MNT="$(readlink -f "$2")"
+
+    sleep 1
+    sudo umount "$MNT"/* || true
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:103]: would it make sense to retry 2 or 3 times instead of sleeping?

File: tests/lib/tools/mkimage-uc22 [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    sleep 1
+    sudo kpartx -d "$IMG" || true
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:105]: same here

File: tests/lib/tools/mkimage-uc22 [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+}
+
+create_image() {
+    IMG="$(readlink -f "$1")"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:109]: local here?

File: tests/lib/tools/mkimage-uc22 (hunk 4 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:103]: I did the change, also removed the '|| true' which is very dangerous

File: tests/lib/tools/mkimage-uc22 (hunk 5 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:105]: done

File: tests/lib/tools/mkimage-uc22 [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+    echo "mkimage-uc22: image assets ready"
+}
+
+# shellcheck disable=SC2317
+cleanup() {
+    local IMG MNT
+    IMG="$(readlink -f "$1")"
+    MNT="$(readlink -f "$2")"
+
+    retry -n 3 --wait 1 sh -c "umount \"$MNT\"/*"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:104]: is `--wait 1` neccessary? isn't the default interval 1s?

File: tests/lib/tools/mkimage-uc22 [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+    retry -n 3 --wait 1 sh -c "kpartx -d \"$IMG\""
+}
+
+create_image() {
+    local IMG
+    IMG="$(readlink -f "$1")"
+
+    rm -f "$IMG"
+    truncate --size=6G "$IMG"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:113]: A comment why we create an image exactly 6G would be nice

File: tests/lib/tools/mkimage-uc22 [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+
+    if [ -d "$CACHE_DIR" ]; then
+        while IFS= read -r mysnap; do
+            echo "$mysnap"
+        done < <(find "$SNAPS_DIR" -name '*.snap')
+    fi
+}
+
+get_assets() {
+    CACHE="$1"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:43]: should CACHE/BASE_* be locals?

File: tests/lib/tools/mkimage-uc22 [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+
+    # rename seed-label
+    mv ./classic-seed/system-seed/systems/"$(date +%Y%m%d)" ./classic-seed/system-seed/systems/"$SEED_LABEL"
+    # and put the seed in place
+    mkdir -p "$DESTDIR"/var/lib/snapd
+    mv ./classic-seed/system-seed "$DESTDIR"/var/lib/snapd/seed
+}
+
+populate_image() {
+    IMG="$1"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:234]: local IMG/CACHE/MNT/MODEL_ASSERTION?

File: tests/lib/tools/mkimage-uc22 [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+    sudo cp -a "$CACHE"/snap-pc-kernel/kernel.efi "$MNT"/ubuntu-boot/EFI/ubuntu/"$KERNEL_SNAP"
+    sudo ln -sf "$KERNEL_SNAP"/kernel.efi "$MNT"/ubuntu-boot/EFI/ubuntu/kernel.efi
+
+    # cleanup
+    sync
+    sudo umount "$MNT"/ubuntu-*
+}
+
+replace_initramfs_bits() {
+    CACHE="$1"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:356]: local CACHE/SNAPD_BINPATH?

File: tests/lib/tools/mkimage-uc22 [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+    objcopy -O binary -j .linux "$KERNEL_EFI_ORIG" linux
+    # Replace kernel.efi in unsquashed snap
+    objcopy --add-section .linux=linux --change-section-vma .linux=0x2000000 \
+            --add-section .initrd=initrd.img.new --change-section-vma .initrd=0x3000000 \
+            usr/lib/ubuntu-core-initramfs/efi/linux*.efi.stub \
+            "$KERNEL_EFI_ORIG"
+}
+
+cleanup() {
+    IMG="$(readlink -f "$1")"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:383]: local IMG/MNT?

File: tests/lib/tools/mkimage-uc22 [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+    MNT="$(readlink -f "$2")"
+
+    sync
+    sleep 1
+    umount "$MNT"/* || true
+    sleep 1
+    kpartx -d "$IMG" || true
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:390]: Maybe use `retry` here ?

File: tests/lib/tools/mkimage-uc22 [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+}
+
+replace_images_files() {
+    IMG="$1"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:394]: local variables here?

File: tests/lib/tools/mkimage-uc22 [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+    # enable debug traces
+    sudo mkdir -p "$MNT"/data/etc/systemd/system/snapd.service.d/
+    sudo tee "$MNT"/data/etc/systemd/system/snapd.service.d/override.conf <<'EOF'
+[Service]
+Environment=SNAPD_DEBUG=1
+EOF
+}
+
+show_how_to_run_qemu() {
+    IMG="$1"
```
Comment (Meulengracht) [inline tests/lib/tools/mkimage-uc22:435]: local?

**Review by Meulengracht — COMMENTED** [review]
Some additional comments, but looks really nice

File: tests/lib/tools/mkimage-uc22 (hunk 7 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:104]: it is not necessary, but I think it helps to understand easily when someone reads the code
The same is for -n3 (which is also the default)

File: tests/lib/tools/mkimage-uc22 (hunk 9 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:43]: yes, added local in all the vars

File: tests/lib/tools/mkimage-uc22 (hunk 10 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:234]: yes

File: tests/lib/tools/mkimage-uc22 (hunk 11 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:356]: done

File: tests/lib/tools/mkimage-uc22 (hunk 12 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:383]: done

File: tests/lib/tools/mkimage-uc22 (hunk 14 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:394]: done

File: tests/lib/tools/mkimage-uc22 (hunk 15 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:435]: done

File: tests/lib/tools/mkimage-uc22 (hunk 13 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:390]: I updated the cleanup function but for some raeson it was repeated, so I deleted this one. Thanks for the comment

File: tests/lib/tools/mkimage-uc22 (hunk 8 above)
Comment (sergiocazzolato) [reply tests/lib/tools/mkimage-uc22:113]: added

**Review by Meulengracht — APPROVED** [review]
Thanks for the changes, lgtm

**Pull request merged**